source/CModuleSystem.cpp

```diff
@@ -0,0 +1,306 @@
+
+using namespace CLEO;
+
+void CModuleSystem::Clear()
+{
+	modules.clear();
+}
+
+const ScriptDataRef CModuleSystem::GetExport(const char* moduleName, const char* exportName)
+{
+	// TODO: get current working dir
+	std::string path = "";
+
+	size_t len = strlen(moduleName);
+	if (path.empty())
+	{
+		path.reserve(len);
+	}
+	else
+	{
+		path.reserve(path.length() + len + 1);
+		path.push_back('/'); // path separator
+	}
+
+	path += moduleName;
+	normalizePath(path);
+
+	auto it = modules.find(path);
+	if (it == modules.end()) // module not loaded yet?
+	{
+		if (!LoadFile(path.c_str()))
+		{
+			return {};
+		}
+
+		// check if available now
+		it = modules.find(path);
+		if (it == modules.end())
+		{
+			return {};
+		}
+	}
+	auto& module = it->second;
+
+	return module.GetExport(exportName);
+}
+
+bool CModuleSystem::LoadFile(const char* path)
+{
+	CModule module;
+
+	std::string normalizedPath(path);
+	normalizePath(normalizedPath);
+
+	if (!module.LoadFromFile(normalizedPath.c_str()))
+	{
+		return false;
+	}
+
+	modules[normalizedPath] = std::move(module); // move to container
+	return true;
+}
+
+bool CModuleSystem::LoadDirectory(const char* path)
+{
+	bool result = true;
+
+	for (auto& it : std::filesystem::recursive_directory_iterator(path))
+	{
+		auto& filePath = it.path();
+		if (filePath.extension() == ".s")
+		{
+			result &= LoadFile(filePath.string().c_str());
+		}
+	}
+
+	return result;
+}
+
+bool CModuleSystem::LoadCleoModules()
+{
+	return LoadDirectory("cleo/cleo_modules"); // TODO: select work dir
+}
+
+bool CModuleSystem::Reload()
+{
+	std::set<std::string> names;
+	for (auto& it : modules)
+	{
+		names.insert(it.second.GetFilepath());
+	}
+
+	Clear();
+
+	bool result = true;
+	for (auto& name : names)
+	{
+		result &= LoadFile(name.c_str());
+	}
+
+	return result;
+}
+
+void CLEO::CModuleSystem::normalizePath(std::string& path)
+{
+	for (char& c : path)
+	{
+		// standarize path separators
+		if (c == '/')
+			c = '\\';
+
+		// lower case
+		c = std::tolower(c);
+	};
+}
+
+void CModuleSystem::CModule::Clear()
+{
+	filepath.clear();
+	data.clear();
+	exports.clear();
+}
+
+const char* CModuleSystem::CModule::GetFilepath() const
+{
+	return filepath.c_str();
+}
+
+bool CModuleSystem::CModule::LoadFromFile(const char* path)
+{
+	Clear();
+
+	std::ifstream file(path, std::ios::binary);
+	if (!file.good())
+	{
+		TRACE("Failed to open module file '%s'", path);
+		return false;
+	}
+
+#pragma warning ( push )
+#pragma warning ( disable: 4838 )
+#pragma warning ( disable: 4309 )
+	const char Header_First_Instruction[] = { 0x02, 0x00, 0x01 }; // jump, param type
+	const char Header_Magic[] = { 0xFF, 0x7F, 0xFE, 0x00, 0x00 }; // Rockstar custom header magic
+	const char Header_Module_Signature[] = { 'E', 'X', 'P', 'T' }; // CLEO's module header signature
+#pragma warning ( pop )
+
+#pragma pack(push, 1)
+	struct
+	{
+		char firstInstruction[3];
+		int jumpAddress;
+		char magic[5];
+		char signature[4];
+		int size;
+	} header;
```

Comment:
header is only first three fields. 
`signature[4]` and `size` belong to a section struct that should be defined separately

Reply:
also as discussed, there could be many sections with different signature and their order is unspecified.

Reply:
Done